File: ios/NetworkingModule.swift
import Foundation

/// TurboModule implementation for native HTTP networking.
/// Provides high-performance HTTP for non-UI API calls like tokenization.
@objc(BoltNetworking)
class NetworkingModule: NSObject {

  @objc
  static func moduleName() -> String {
    return "BoltNetworking"
  }

  @objc
  static func requiresMainQueueSetup() -> Bool {
    return false
  }

  @objc
  func request(_ method: String,
               url: String,
               headers: String,
               body: String,
               resolve: @escaping (Any) -> Void,
               reject: @escaping (String, String, NSError) -> Void) {
    guard let requestUrl = URL(string: url) else {
      reject("INVALID_URL", "Invalid URL: \(url)", NSError(domain: "BoltNetworking", code: 1))
      return
    }

    var request = URLRequest(url: requestUrl)
    request.httpMethod = method

    // Parse and set headers
    if let headersData = headers.data(using: .utf8),
       let headerDict = try? JSONSerialization.jsonObject(with: headersData) as? [String: String] {
      for (key, value) in headerDict {
        request.setValue(value, forHTTPHeaderField: key)
      }
    }

    // Set body if not empty
    if !body.isEmpty {
      request.httpBody = body.data(using: .utf8)
    }

    URLSession.shared.dataTask(with: request) { data, response, error in
      if let error = error {
        reject("NETWORK_ERROR", error.localizedDescription, error as NSError)
        return
      }

      let httpResponse = response as? HTTPURLResponse
      let status = httpResponse?.statusCode ?? 0
      let responseHeaders = httpResponse?.allHeaderFields as? [String: String] ?? [:]
      let responseBody = data.flatMap { String(data: $0, encoding: .utf8) } ?? ""

      let result: [String: Any] = [
        "status": status,
        "headers": responseHeaders,
        "body": responseBody
      ]

      if let jsonData = try? JSONSerialization.data(withJSONObject: result),
         let jsonString = String(data: jsonData, encoding: .utf8) {
        resolve(jsonString)
      } else {
        reject("SERIALIZE_ERROR", "Failed to serialize response", NSError(domain: "BoltNetworking", code: 2))
      }
    }.resume()
  }
}

File: android/src/main/java/com/boltreactnativesdk/NetworkingModule.kt
package com.boltreactnativesdk

import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import org.json.JSONObject
import java.io.BufferedReader
import java.io.InputStreamReader
import java.io.OutputStreamWriter
import java.net.HttpURLConnection
import java.net.URL

/**
 * TurboModule implementation for native HTTP networking on Android.
 * Provides high-performance HTTP for non-UI API calls like tokenization.
 */
class NetworkingModule(reactContext: ReactApplicationContext) :
    ReactContextBaseJavaModule(reactContext) {

    companion object {
        const val NAME = "BoltNetworking"
    }

    override fun getName(): String = NAME

    @ReactMethod
    fun request(method: String, url: String, headers: String, body: String, promise: Promise) {
        Thread {
            try {
                val connection = URL(url).openConnection() as HttpURLConnection
                connection.requestMethod = method

                // Parse and set headers
                try {
                    val headerObj = JSONObject(headers)
                    val keys = headerObj.keys()
                    while (keys.hasNext()) {
                        val key = keys.next()
                        connection.setRequestProperty(key, headerObj.getString(key))
                    }
                } catch (_: Exception) {
                    // headers may be empty or invalid JSON
                }

                // Set body if not empty
                if (body.isNotEmpty() && method != "GET" && method != "HEAD") {
                    connection.doOutput = true
                    OutputStreamWriter(connection.outputStream).use { writer ->
                        writer.write(body)
                        writer.flush()
                    }
                }

                val status = connection.responseCode
                val responseHeaders = JSONObject()
                connection.headerFields?.forEach { (key, values) ->
                    if (key != null && values.isNotEmpty()) {
                        responseHeaders.put(key, values.joinToString(", "))
                    }
                }

                val inputStream = if (status >= 400) {
                    connection.errorStream
                } else {
                    connection.inputStream
                }

                val responseBody = inputStream?.let {
                    BufferedReader(InputStreamReader(it)).readText()
                } ?: ""

                val result = JSONObject()
                result.put("status", status)
                result.put("headers", responseHeaders)
                result.put("body", responseBody)

                promise.resolve(result.toString())

                connection.disconnect()
            } catch (e: Exception) {
                promise.reject("NETWORK_ERROR", e.message, e)
            }
        }.start()
    }
}
